Misra fixes

ceolin opened the pull request.

Commit:
logging: Simplify function to get number of log backends

As we subtracting pointers of the same type, it is not necessary cast
it to u8_t. This simplify and avoid some MISRA-C violations, e.g rule
11.8.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
kernel: Use NULL instead of 0

MISRA-C rule 11.9

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
sys_clock: Explicitly set constants sign

Explicitly setting these constants as unsigned.

MISRA rule 10.8

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
sys_clock: Change function signature

__ticks_to_ms always return an unsigned value, changing the function
to reflect it.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
kernel: Change function signature

Change k_timer_remaining_get to return unsigned.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
arch: x86: Use macro BIT to shift bits

The operation was shifiting bit using a signed constant in the left
operand. Use BIT macro to do it properly.

MISRA-C rule 12.2

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

andrewboie commented:
code changes look OK, pls fix commit message to have a body

ceolin commented:
done :)

ceolin commented:
I'll check the failures, I guess they are related with print format

andyross commented on include/sys_clock.h:
```diff
@@ -125,23 +125,21 @@ static ALWAYS_INLINE s32_t _ms_to_ticks(s32_t ms)
-	u32_t ms_per_tick = MSEC_PER_SEC / CONFIG_SYS_CLOCK_TICKS_PER_SEC;
-
-	return (u64_t)ticks * ms_per_tick;
+	return (u64_t)ticks * (MSEC_PER_SEC / (u64_t)CONFIG_SYS_CLOCK_TICKS_PER_SEC);
```
Same bug

andyross commented on include/sys_clock.h:
```diff
@@ -125,23 +125,21 @@ static ALWAYS_INLINE s32_t _ms_to_ticks(s32_t ms)
 #ifdef _NEED_PRECISE_TICK_MS_CONVERSION
 	/* use 64-bit math to keep precision */
-	return (u64_t)ticks * MSEC_PER_SEC / CONFIG_SYS_CLOCK_TICKS_PER_SEC;
+	return (u64_t)ticks * (MSEC_PER_SEC / (u64_t)CONFIG_SYS_CLOCK_TICKS_PER_SEC);
```
No, this is a precision bug.  The multiplication has to happen first, because the division will compute something along the lines of 1000/100000000 and produce a zero.

What's the MISRA rule being addressed here?  Just that expressions must be fully parenthesized?  The * and / operators are left-to-right associative and have the same precedence.

ceolin replied:
indeed, left to right evaluation :/
I'll change it.

andyross commented on kernel/thread.c:
```diff
@@ -100,9 +100,9 @@ void _impl_k_busy_wait(u32_t usec_to_wait)
-		(u64_t)USEC_PER_SEC
+		usec_to_wait *
+		sys_clock_hw_cycles_per_sec() /
+		USEC_PER_SEC
```
This looks dangerous too.  Presumably those casts were there to guarantee that the math was done in 64 bit precision.  So now that that multiply is going to be done in 32 bits and potentially overflow where it didn't before.

ceolin replied:
Initially I was changing I was changing the constant to be long long. Then other problems appeared, I've undone the changes and forgot it.
The MISRA-C violation is promote a 32 wide bit variable to a 64.

andyross reviewed (changes requested):
Looks like the math semantics got changed in a few places by mistake.

Commit:
tests: benchmark: Remove redundant define

The constant NSEC_PER_USEC is already defined in sys_clock.h, there is
no need to define it here.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

andyross commented on kernel/thread.c:
```diff
@@ -99,10 +99,11 @@ void _impl_k_busy_wait(u32_t usec_to_wait)
-		(u64_t)USEC_PER_SEC
+		usec_to_wait *
+		sys_clock_hw_cycles_per_sec() /
+		usecs
```
This still looks wrong to me.  That is going to evaluate as:

    (usec_to_wait * s_c_h_c_p_s() )  / usecs

The multiplication is still being done in 32 bits and will overflow where the original expression wasn't.

I also still don't understant why MISRA disallows doing math in higher precision.  Does it issue the same warning if you want to extract a bitfield or take a char variable and use it as a multiplier on an int?

ceolin replied:
arghh, indeed. I'll remove this commit from this pr, since it will require some extra work.

ceolin commented:
@andyross I've removed the problematic commit, the rest is ok ?

ceolin commented:
@andyross ping

andyross reviewed (approved):
Can't see anything wrong.  Still not clear on exactly what the rule being addressed says, but the casts in these expressions aren't wrong or unclear.
